src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -0,0 +1,154 @@
+            {
+                var changes = result.Edits.Select(e => e.AsTextChange(originalText)).ToArray();
+                changedText = changedText.WithChanges(changes);
+                changedContext = await context.WithTextAsync(changedText);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Format the razor bits of the file
+            var syntaxTree = changedContext.CodeDocument.GetSyntaxTree();
+            var edits = FormatRazor(changedContext, syntaxTree);
+
+            // Compute the final combined set of edits
+            var formattingChanges = edits.Select(e => e.AsTextChange(changedText));
+            changedText = changedText.WithChanges(formattingChanges);
+            var finalChanges = SourceTextDiffer.GetMinimalTextChanges(originalText, changedText, lineDiffOnly: false);
+            var finalEdits = finalChanges.Select(f => f.AsTextEdit(originalText)).ToArray();
+
+            return new FormattingResult(finalEdits);
+        }
+
+        private IEnumerable<TextEdit> FormatRazor(FormattingContext context, RazorSyntaxTree syntaxTree)
+        {
+            var edits = new List<TextEdit>();
+            var source = syntaxTree.Source;
+
+            foreach (var node in syntaxTree.Root.DescendantNodes())
+            {
+                // Disclaimer: CSharpCodeBlockSyntax is used a _lot_ in razor so I'm being overly careful to only try to
+                // format syntax forms we care about.
+                //
+                // We're looking for a code block like this:
+                //
+                // @code {
+                //    var x = 1;
+                // }
+                //
+                // The nodes will be a grandchild of a RazorDirective (the "@code") and we expect there to be
+                // at least three children, being:
+                // 1. Optional whitespace
+                // 2. The opening brace
+                // 3. The C# code
+                // 4. The closing brace
+                if (node is CSharpCodeBlockSyntax code &&
+                    node.Parent?.Parent is RazorDirectiveSyntax directive &&
+                    directive.DirectiveDescriptor?.Kind == DirectiveKind.CodeBlock)
+                {
+                    var children = code.Children;
+                    if (children.Count == 4 && code.Children[0] is UnclassifiedTextLiteralSyntax literal)
+                    {
+                        // For whitespace we normalize it differently depending on if its multi-line or not
+                        FormatWhitespaceBetweenDirectiveAndBrace(literal, directive, edits, source, context);
+                    }
+                    else if (children.Count == 3 && children[0] is RazorMetaCodeSyntax metaCode)
+                    {
+                        var metas = metaCode.MetaCode;
+                        if (metas.Count == 1 &&
+                            metas[0].Content.Equals("{"))
```

Comment:
This bit and the `children` bit above would be much nicer with C# 9 patterns.. just sayin' 😛